src/app/Library/CrudPanel/Traits/Fields.php

```diff
@@ -35,6 +35,17 @@ public function addField($field)
             $newField = $field;
         }
 
+        //we set this up exclusive for relationship type field. in future we hope we can figure out this automatically from field name
+        //atm we avoid any breaking changes while developing the guessing habilities for crud panel
+        if (isset($newField['type']) && $newField['type'] == 'relationship') {
```

Comment:
Ugh... Doing something in the ```addField()``` method depending on the field ```type``` is something I've avoided for 3 years 😆 Because it means other field _types_ won't be able to do the same thing - this would be strictly for that field. However 😁  Since this field is supposed to solve all _our relationship problems_ (pun intended), I think it would be a trade worth doing. _However_. I may have a solution for this. **What if, instead of checking if it's the relationship field, we check that it has ```entity```, but it doesn't have everything else that we need for a relationship field (```model``` and ```pivot``` afaik)?**

So:
- first we check if the ```type``` is set
- if not, we see if the ```type``` should be something non-relationship (infer from the db column)
- if not, we see if there should be an ```entity```; if so, set both ```entity``` and ```type``` to ```relationship```;
- if ```entity``` is set, but no ```model``` or ```pivot```, we determine ```model``` and ```pivot```;

Something _rough_ like this should work:

```diff

    /**
     * Add a field to the create/update form or both.
     *
     * @param string|array $field The new field.
     *
     * @return self
     */
    public function addField($field)
    {
        // if the field_definition_array array is a string, it means the programmer was lazy and has only passed the name
        // set some default values, so the field will still work
        if (is_string($field)) {
            $newField['name'] = $field;
        } else {
            $newField = $field;
        }

-        // we set this up exclusive for relationship type field
-        // atm we avoid any breaking changes while developing the guessing abilities for crud panel
-        if (isset($newField['type']) && $newField['type'] == 'relationship') {
-            $relationData = $this->getRelationFromFieldName($newField['name']);
-            if ($relationData) {
-                $newField = array_merge($newField, $relationData);
-            } else {
-                abort(500, 'Unable to process relationship field: '.$field['name']);
-            }
-        }

        // if this is a relation type field and no corresponding model was specified, 
        // get it from the relation method defined in the main model
        if (isset($newField['entity']) && ! isset($newField['model'])) {
            $newField['model'] = $this->getRelationModel($newField['entity']);
        }

        // if the label is missing, we should set it
        if (! isset($newField['label'])) {
            $label = is_array($newField['name']) ? $newField['name'][0] : $newField['name'];
            $newField['label'] = mb_ucfirst(str_replace('_', ' ', $label));
        }

        // if the field type is missing, we should set it
+       // first, try to infer the FIELD type from the DB COLUMN type
+       // but only if there's actually a db column type for this field 
-        if (! isset($newField['type'])) {
+       if (! isset($newField['type']) && in_array($newField['name'], $this->getDbColumnsNames())) {
               $newField['type'] = $this->getFieldTypeFromDbColumnType($newField['name']);
        }

+        // second, by this point, if the field type is STILL missing, 
+        // it means there's no db column with this name
+        // so let's check if there's a relationship on the model with this name
+       if (! isset($newField['type'])) {
+           // we specifically call getRelationFromFieldName() only if no type is already selected
+           // because it's a very resource-intensive operation 
+           // (there are A LOT of methods on an Eloquent model) 
+           $relationData = $this->getRelationFromFieldName($newField['name']);
+
+           if ($relationData) {
+                $newField = array_merge($newField, $relationData);
+                $newField['type'] = 'relationship';
+           }
+       }

+        // if this is a relation type field, but the model STILL hasn't been determined, let's do that now
+        // this block will be called, for example, if the user has specified "entity", "model" but no "pivot"
+        if (isset($newField['entity']) && (! isset($newField['model']) || !isset($newField['pivot'])))) {
+            $relationData = $this->getRelationFromFieldName($newField['name']);
+            if ($relationData) {
+                $newField = array_merge($newField, $relationData);
+                $newField['type'] = 'relationship';
+            } else {
+                abort(500, 'Unable to process relationship field: '.$field['name']);
+            }
+        }


        // if a tab was mentioned, we should enable it
        if (isset($newField['tab'])) {
            if (! $this->tabsEnabled()) {
                $this->enableTabs();
            }
        }

        $fields = $this->getOperationSetting('fields');
        $fieldKey = is_array($newField['name']) ? implode('_', $newField['name']) : $newField['name'];
        $fields = array_add($this->fields(), $fieldKey, $newField);
        $this->setOperationSetting('fields', $fields);

        return $this;
    }
```

Now... there's downsides to this, of course, as with anything.

PRO:
- we no longer determine the field ```model``` and ```pivot``` _just_ for the "relationship" field; so now other fields could benefit from this too, fields that previously could infer the ```model```, but not ```pivot```; so for fields like ```fetch```, ```select```, ```select2_multiple```, ```select2_from_ajax```, ```select2_from_ajax_multiple```, the definition would not need to include ```pivot``` anymore, since that's inferred; which is cool;

CON:
- quite a bit more complicated 😄 Of course, all this new code can be extracted in a separate method, something like 
    - ```inferFieldType($field)``` that itself calls ```getFieldTypeFromDbColumnType()```
    - ```inferRelationshipDataForField(array $field)```, so it's easier to read;

In the end, after refactoring, ```addField()``` could look something like this:

```php
public function addField($field) {

        $newField = $this->inferFieldName($newField);
        $newField = $this->inferFieldLabel($newField);
        $newField = $this->inferFieldType($newField);
        $newField = $this->inferFieldEntity($newField);
        $newField = $this->inferFieldModelAndPivot($newField);

        $arrayKey = $this->getArrayKeyFromFieldName($field);
        $this->addFieldToCurrentOperation($arrayKey, $newField);

        return $this;
}
```

**What do you think about this @pxpm ?** I'll do this myself, I think it's faster that way, but I could use your opinion - is it something that's _better_, or just _different_? 😄

Reply:
You are right of course, i also did not agree 100% having exclusive code for this field type and I suspected that you would also not be at this party that's why I left the comment.

Why I started with this field only? Because when using it, we will override any previous values set by dev. 
When using your example, if I specify a model and don't specify pivot, we will infer the info from the relation and override any developer values in in specific keys like: `model,entity`. Because we grab them from the relation definition, and that sould be the only source of truth, in these fields.

About the refactor I think what you purposed is the way to go, also it would allows us to write test for those new functions and have a better separarion of concerns. I would go with your suggestion.

Reply:
Great to hear that Pedro! I'm on it.

> Why I started with this field only? Because when using it, we will override any previous values set by dev.
> When using your example, if I specify a model and don't specify pivot, we will infer the info from the relation and override any developer values in in specific keys like: model,entity. Because we grab them from the relation definition, and that sould be the only source of truth, in these fields.

Hmm... I'm afraid I don't agree here. Sure, the information we infer from the relationship is _most likely_ the best. But if the developer _chooses to specify_ something, that must overwrite the default. We have to assume we know what he's doing, if he specifies something. That's the way it works in Backpack everywhere - the developer is king! If he wants to overwrite something, Backpack shouldn't change his value, in my opinion...

Reply:
Ok it's DONE! Not only that, but splitting the ```addField()``` method into multiple smaller methods made the code quality service rate ```Fields.php``` as "Grade A" instead of "Grade F" 🎉 🎈  Which... you know... matters for some people 😄 Not me... but some people 😆 

Bonus: we also detect "entity" now. So if you do ```$this->crud->addField("category_id")``` you now get a relationship field ;-)

There's one catch. Because of the choice that attributes specified by the developer NEVER get overwritten, one use case @pxpm  coded no longer works. Previously if you did:

```php
$this->crud->addField([
   'name' => 'category',
   'entity' => 'category',
]);
```
Backpack would _overwrite_ the name from ```category``` to ```category_id```. Now it doesn't do that anymore. So the syntax stays the same as it is for select2 etc:
```php
$this->crud->addField([
   'name' => 'category_id',
   'entity' => 'category',
]);
```

Basically, when adding relationship fields, the name needs to be:
- 1-n relationship - the foreign key
- n-n relationship - the name of the relationship

This is good, because it's now backwards-compatible with the old relationship fields.
But it's also bad, because there's a different way to define that field depending on the relationship type...

---

We _could_ choose to:
- change ```makeSureFieldHasEntity()``` to overwrite the ```name``` for 1-n relationships, if the name doesn't match the foreign key;
- revert to how @pxpm did it, and overwrite _everything_ that we infer from the relationship;

Food for thought. Feedback appreciated.